### PR TITLE
Fixes for broken Url's

### LINKED ISSUE #1230: [Regression] wikisource en_all_maxi only scraps first page and a couple of images
With the two latest iterations we end up with a [25Mb file](https://farm.openzim.org/pipeline/5f17ef3be2aaee590569667d) instead of [20 Gb](https://farm.openzim.org/pipeline/5ed3f0a658b7a2d9e72630a9).

### PR DESCRIPTION
**Pull request by bakshiutkarsha**
should fix #1230 and #1203

### CODE REVIEW

**Comment by bakshiutkarsha** [comment]
Summary of the PR:

For wikisource ApiUrl:
`https://en.wikisource.org/api/rest_v1/page/html/articleID`

For bm.wikipedia, ApiUrl:
`https://bm.wikipedia.org/api/rest_v1/page/mobile-sections/articleID`

For gbf.wiki, ApiUrl:
`https://gbf.wiki/api.php?action=visualeditor&mobileformat=html&format=json&paction=parse&page=articleID`
And now with correct API, the internal links are also correct, hence issue #1203 is also fixed

**Comment by kelson42** [comment]
You need a test using local parsoid as well.

File: src/Downloader.ts [hunk 1]
```diff
@@ -184,37 +184,56 @@ class Downloader {
     return `${cachedPart}${path}`;
   }
 
-  public async checkCapabilities(): Promise<void> {
-    // check if RESTBase-powered API available
+  public setBaseUrls(baseUrl: string, baseUrlForMainPage: string = this.mw.veApiUrl.href){
```
Comment (kelson42) [inline src/Downloader.ts:187]: This function should take no argument. Values should be set based on `this.mw` and `this.mwCapabilities`.

File: src/Downloader.ts [hunk 2]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
+    this.baseUrl = baseUrl;
+    this.baseUrlForMainPage = baseUrlForMainPage;
+  }
+
+  public async checkRestApiAvailabilty(type: string, restApiBase: string, defaultUrlForMainPage: string){
     try {
-      const restApiMainPageQuery = await this.getJSON<any>(`${this.baseUrl}${encodeURIComponent(this.mw.metaData.mainPage)}`);
-      this.mwCapabilities.restApiAvailable = !!restApiMainPageQuery.lead;
+      const restBaseApiQuery = await this.getJSON<any>(`${restApiBase}${type}${encodeURIComponent(this.mw.metaData.mainPage)}`);
+      this.setBaseUrls(`${restApiBase}${type}`);
```
Comment (kelson42) [inline src/Downloader.ts:195]: Not this should be called outside `checkRestApiAvailabilty` (probably just after).

File: src/Downloader.ts [hunk 3]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
-      if (!this.mwCapabilities.restApiAvailable || !this.mwCapabilities.veApiAvailable) {
-        logger.log(`Using local MCS and ${this.mwCapabilities.veApiAvailable ? 'remote' : 'local'} Parsoid`);
+      if (!this.mwCapabilities.veApiAvailable) {
+        logger.log(`Using local MCS and ${this.mwCapabilities.veApiAvailable} local Parsoid`);
         await this.initLocalServices();
 
-        if (!this.mwCapabilities.restApiAvailable) {
-          this.baseUrl = `http://localhost:6927/${this.mw.webUrl.hostname}/v1/page/mobile-sections/`;
+        if (!this.mwCapabilities.mobileRestApiAvailable) {
+          this.setBaseUrls(`http://localhost:6927/${this.mw.webUrl.hostname}/v1/page/mobile-sections/`)
```
Comment (kelson42) [inline src/Downloader.ts:232]: No, `checkCapabilities` should only set `this.mwCapabilities`. Base URLs should be set by `setBaseUrls()` (called afterward).

File: src/util/articleRenderers.ts [hunk 4]
```diff
@@ -54,7 +52,10 @@ export const renderArticle = async (json: any, articleId: string, dump: Dump, ca
             await articleDetailXId.set(_articleId, _articleDetail);
         }
 
-        const html = renderMCSArticle(json, dump, _articleId, _articleDetail);
+        if(capabilities.mobileRestApiAvailable || !capabilities.desktopRestApiAvailable){
```
Comment (kelson42) [inline src/util/articleRenderers.ts:55]: Why testing `!capabilities.desktopRestApiAvailable` ? Does not make sense to me.

File: src/Downloader.ts [hunk 5]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
+    await this.checkRestApiAvailabilty('html/', restApiBase, defaultUrlForMainPage);
+    await this.checkRestApiAvailabilty('mobile-sections/', restApiBase, defaultUrlForMainPage);
 
     if (!this.forceLocalParser) {
       // check if VisualEditor available
       try {
         const parsoidMainPageQuery = await this.getJSON<any>(`${this.mw.veApiUrl.href}${encodeURIComponent(this.mw.metaData.mainPage)}`);
         this.mwCapabilities.veApiAvailable = !!parsoidMainPageQuery.visualeditor.content;
+
+        if(!this.mwCapabilities.desktopRestApiAvailable && !this.mwCapabilities.mobileRestApiAvailable){
```
Comment (kelson42) [inline src/Downloader.ts:217]: This should not be there. Only set `this.mwCapabilities`.

File: src/Downloader.ts [hunk 6]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
         this.mwCapabilities.veApiAvailable = false;
         logger.warn(`Failed to get remote Parsoid`);
       }
     }
 
     if (!this.noLocalParserFallback) {
-      if (!this.mwCapabilities.restApiAvailable || !this.mwCapabilities.veApiAvailable) {
-        logger.log(`Using local MCS and ${this.mwCapabilities.veApiAvailable ? 'remote' : 'local'} Parsoid`);
+      if (!this.mwCapabilities.veApiAvailable) {
+        logger.log(`Using local MCS and ${this.mwCapabilities.veApiAvailable} local Parsoid`);
```
Comment (kelson42) [inline src/Downloader.ts:228]: This log entry should not be there but in `setBaseUrls()` which is the part which decides which API (mobilerest, desktoprest, ve, parsoid,... ) we are using.

File: src/Downloader.ts [hunk 7]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
         await this.initLocalServices();
```
Comment (kelson42) [inline src/Downloader.ts:229]: This should be as well done outside `checkCapabilities()`

File: src/Downloader.ts [hunk 8]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
 
-        if (!this.mwCapabilities.restApiAvailable) {
-          this.baseUrl = `http://localhost:6927/${this.mw.webUrl.hostname}/v1/page/mobile-sections/`;
+        if (!this.mwCapabilities.mobileRestApiAvailable) {
+          this.setBaseUrls(`http://localhost:6927/${this.mw.webUrl.hostname}/v1/page/mobile-sections/`)
         }
+
         if (!this.mwCapabilities.veApiAvailable) {
-          this.baseUrlForMainPage = `http://localhost:8000/${this.mw.webUrl.hostname}/v3/page/pagebundle/`;
+          this.setBaseUrls(`${restApiBase}mobile-sections`, `http://localhost:8000/${this.mw.webUrl.hostname}/v3/page/pagebundle/`)
```
Comment (kelson42) [inline src/Downloader.ts:236]: Same, not here.

File: src/Downloader.ts [hunk 9]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
       try {
         const parsoidMainPageQuery = await this.getJSON<any>(`${this.mw.veApiUrl.href}${encodeURIComponent(this.mw.metaData.mainPage)}`);
         this.mwCapabilities.veApiAvailable = !!parsoidMainPageQuery.visualeditor.content;
+
+        if(!this.mwCapabilities.desktopRestApiAvailable && !this.mwCapabilities.mobileRestApiAvailable){
+          this.setBaseUrls(defaultUrlForMainPage);
+        }
       } catch (err) {
         this.mwCapabilities.veApiAvailable = false;
         logger.warn(`Failed to get remote Parsoid`);
```
Comment (kelson42) [inline src/Downloader.ts:222]: We should probably have a global info log entry informing about how the capabilities look like at the end of the method.

File: src/Downloader.ts [hunk 10]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
-      this.mwCapabilities.restApiAvailable = false;
-      logger.warn(`Failed to get remote Rest API`);
+      (type === 'mobile-sections/') ?  this.mwCapabilities.mobileRestApiAvailable = false : this.mwCapabilities.desktopRestApiAvailable = false;
+      logger.warn(`Failed to get remote ${type} API`);
     }
+  }
+
+  public async checkCapabilities(): Promise<void> {
+    // check if RESTBase-powered API available
+    const restApiBase = `${this.mw.restApiUrl.href}page/`;
```
Comment (kelson42) [inline src/Downloader.ts:205]: All the URLs should come from `this.mw`. We should have a `this.mw.mobileRestApiUrl`... etc. I don't want any hardcoded URL at this level.

File: src/Downloader.ts [hunk 11]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
   }
 
-  public async checkCapabilities(): Promise<void> {
-    // check if RESTBase-powered API available
+  public setBaseUrls(baseUrl: string, baseUrlForMainPage: string = this.mw.veApiUrl.href){
+    this.baseUrl = baseUrl;
+    this.baseUrlForMainPage = baseUrlForMainPage;
+  }
+
+  public async checkRestApiAvailabilty(type: string, restApiBase: string, defaultUrlForMainPage: string){
```
Comment (kelson42) [inline src/Downloader.ts:192]: I believe you can make this method more generic (based on HTTP response status code), returning a boolean, called `checkApiAvailability(url)`

File: src/Downloader.ts [hunk 12]
```diff
@@ -184,37 +184,56 @@ class Downloader {
...
+    // check if RESTBase-powered API available
+    const restApiBase = `${this.mw.restApiUrl.href}page/`;
+    const defaultUrlForMainPage = this.mw.veApiUrl.href;
+
+    await this.checkRestApiAvailabilty('html/', restApiBase, defaultUrlForMainPage);
+    await this.checkRestApiAvailabilty('mobile-sections/', restApiBase, defaultUrlForMainPage);
 
     if (!this.forceLocalParser) {
       // check if VisualEditor available
       try {
```
Comment (kelson42) [inline src/Downloader.ts:213]: You should use `checkApiAvailability(url)`.

File: src/Downloader.ts [hunk 13]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
 
-  public async checkCapabilities(): Promise<void> {
-    // check if RESTBase-powered API available
-    try {
-      const restApiMainPageQuery = await this.getJSON<any>(`${this.baseUrl}${encodeURIComponent(this.mw.metaData.mainPage)}`);
-      this.mwCapabilities.restApiAvailable = !!restApiMainPageQuery.lead;
-    } catch (err) {
-      this.mwCapabilities.restApiAvailable = false;
-      logger.warn(`Failed to get remote Rest API`);
+  public async setBaseUrls(){
```
Comment (kelson42) [inline src/Downloader.ts]: Missing space here.

File: src/Downloader.ts [hunk 14]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
+    const isVeApiAvailable = this.mwCapabilities.veApiAvailable;
+    const isMobileApiAvailable = this.mwCapabilities.mobileRestApiAvailable;
+    const isDesktopApiAvailable = this.mwCapabilities.desktopRestApiAvailable;
+
+    // Setting the default values
+    this.baseUrlForMainPage = this.mw.veApiUrl.href;
+    this.baseUrl = this.mw.veApiUrl.href;
+
+    // Check for desktop API
+    if(isDesktopApiAvailable){
```
Comment (kelson42) [inline src/Downloader.ts]: missing space here

File: src/Downloader.ts [hunk 15]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
-    try {
-      const restApiMainPageQuery = await this.getJSON<any>(`${this.baseUrl}${encodeURIComponent(this.mw.metaData.mainPage)}`);
-      this.mwCapabilities.restApiAvailable = !!restApiMainPageQuery.lead;
-    } catch (err) {
-      this.mwCapabilities.restApiAvailable = false;
-      logger.warn(`Failed to get remote Rest API`);
+  public async setBaseUrls(){
+    const isVeApiAvailable = this.mwCapabilities.veApiAvailable;
+    const isMobileApiAvailable = this.mwCapabilities.mobileRestApiAvailable;
+    const isDesktopApiAvailable = this.mwCapabilities.desktopRestApiAvailable;
```
Comment (kelson42) [inline src/Downloader.ts]: I don't believe we need this variable, use directly values in `this.mwCapabilities`.

File: src/Downloader.ts [hunk 16]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
-      // check if VisualEditor available
-      try {
-        const parsoidMainPageQuery = await this.getJSON<any>(`${this.mw.veApiUrl.href}${encodeURIComponent(this.mw.metaData.mainPage)}`);
-        this.mwCapabilities.veApiAvailable = !!parsoidMainPageQuery.visualeditor.content;
-      } catch (err) {
-        this.mwCapabilities.veApiAvailable = false;
-        logger.warn(`Failed to get remote Parsoid`);
-      }
+    // Check for Mobile API
+    if(isMobileApiAvailable){
```
Comment (kelson42) [inline src/Downloader.ts]: missing space

**Review by kelson42 — CHANGES_REQUESTED** [review]

**Comment by kelson42** [comment]
@bakshiutkarsha I think I have erased the branch with an older version. Please force push your local branch.

File: src/Downloader.ts (hunk 15 above)
Comment (bakshiutkarsha) [reply src/Downloader.ts]: I tried that as well, but that is making the code really hard to read and that is why I switched to usage of variables, for clean code and better understandability.

File: src/Downloader.ts [hunk 17]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
+    const isMobileApiAvailable = this.mwCapabilities.mobileRestApiAvailable;
+    const isDesktopApiAvailable = this.mwCapabilities.desktopRestApiAvailable;
+
+    // Setting the default values
+    this.baseUrlForMainPage = this.mw.veApiUrl.href;
+    this.baseUrl = this.mw.veApiUrl.href;
+
+    // Check for desktop API
+    if(isDesktopApiAvailable) {
+      this.baseUrl = this.mw.desktopRestApiUrl.href;
```
Comment (kelson42) [inline src/Downloader.ts:196]: Use a ternary operator

File: src/Downloader.ts [hunk 18]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
-      this.mwCapabilities.restApiAvailable = false;
-      logger.warn(`Failed to get remote Rest API`);
+  public async setBaseUrls() {
+    const isVeApiAvailable = this.mwCapabilities.veApiAvailable;
+    const isMobileApiAvailable = this.mwCapabilities.mobileRestApiAvailable;
+    const isDesktopApiAvailable = this.mwCapabilities.desktopRestApiAvailable;
+
+    // Setting the default values
+    this.baseUrlForMainPage = this.mw.veApiUrl.href;
+    this.baseUrl = this.mw.veApiUrl.href;
```
Comment (kelson42) [inline src/Downloader.ts:192]: Here a ternary operator with `isMobileApiAvailable` test

File: src/Downloader.ts [hunk 19]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
 
-  public async checkCapabilities(): Promise<void> {
-    // check if RESTBase-powered API available
-    try {
-      const restApiMainPageQuery = await this.getJSON<any>(`${this.baseUrl}${encodeURIComponent(this.mw.metaData.mainPage)}`);
-      this.mwCapabilities.restApiAvailable = !!restApiMainPageQuery.lead;
-    } catch (err) {
-      this.mwCapabilities.restApiAvailable = false;
-      logger.warn(`Failed to get remote Rest API`);
+  public async setBaseUrls() {
```
Comment (kelson42) [inline src/Downloader.ts:185]: This function as still many problems:
* `isMobileApiAvailable` should be `isMobileRestApiAvailable`
* `isDesktopApiAvailable` should be `isDesktopRestApiAvailable`
* `isVeApiAvailable` is never used (but it should)
* Please do one logic bloc with the necessary `if` and `else` to set `this.baseUrl` ... and an other one to set `this.baseUrlForMainPage`. That way each logic is easly understandable and I don't have to read the whole function code.

File: src/Downloader.ts [hunk 20]
```diff
@@ -184,45 +182,61 @@ class Downloader {
...
-      logger.log(`Using remote MCS/Parsoid`);
     }
 
+    logger.log('Base Url: ', this.baseUrl, '\n', 'Base Url for Main Page: ', this.baseUrlForMainPage)
+  }
+
+  public async checkApiAvailabilty(url: string): Promise<boolean>{
+    try {
+      const apiResponse = await this.getJSON<any>(`${url}${encodeURIComponent(this.mw.metaData.mainPage)}`);
+      return !!apiResponse || !!apiResponse.lead || !!apiResponse.visualeditor.content;
```
Comment (kelson42) [inline src/Downloader.ts:228]: You don't rely on the HTTP status code. Could you please tell me which URL returns a HTTP 200 code even if the API is not available?

**Review by kelson42 — CHANGES_REQUESTED** [review]
@bakshiutkarsha I have requested a few changes in `setBaseUrls()`, but the rest looks pretty OK now.

**Comment by kelson42** [comment]
@bakshiutkarsha Are you ready for a new review?

**Comment by kelson42** [comment]
@bakshiutkarsha I have fixed the piece of code which was problematic. Would you be able please to secure this is good for you and fix the automated tests? Then we could merge.

**Comment by kelson42** [comment]
@bakshiutkarsha I have remarked that the problem was outside the code, so I'm going to merge to `master`

**Pull request merged**